8316681: Rewrite URLEncoder.encode to use small reusable buffers

cl4es opened the pull request:
`URLEncoder` currently appends chars that needs encoding into a `java.io.CharArrayWriter`, converts that to a `String`, uses `String::getBytes` to get the encoded bytes and then appends these bytes in a escaped manner to the output stream. This is somewhat inefficient.

This PR replaces the `CharArrayWriter` with a reusable `CharBuffer` + `ByteBuffer` pair. This allows us to encode to the output `StringBuilder` in small chunks, with greatly reduced allocation as a result.

The exact size of the buffers is an open question, but generally it seems that a tiny buffer wins by virtue of allocating less, and that the per chunk overheads are relatively small.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316681](https://bugs.openjdk.org/browse/JDK-8316681): Rewrite URLEncoder.encode to use small reusable buffers (**Enhancement** - P4)


### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [a578f02d](https://git.openjdk.org/jdk/pull/15865/files/a578f02d4cf152a11bf35cd271f463894112e4a4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15865/head:pull/15865` \
`$ git checkout pull/15865`

Update a local copy of the PR: \
`$ git checkout pull/15865` \
`$ git pull https://git.openjdk.org/jdk.git pull/15865/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15865`

View PR using the GUI difftool: \
`$ git pr show -t 15865`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15865.diff">https://git.openjdk.org/jdk/pull/15865.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15865#issuecomment-1729812315)

cl4es commented:
Micros show a small throughput win and a large allocation reduction:
```
Name                             (unchanged) Cnt        Base    Error         Test    Error   Unit   Diff%
URLEncodeDecode.testEncodeLatin1           0  15       3.471 ±  0.103        2.915 ±  0.044  ms/op   16.0% (p = 0.000*)
  :gc.alloc.rate.norm                    N/A  15 3013680.062 ±  0.721  1964156.209 ±  0.308   B/op  -34.8% (p = 0.000*)
URLEncodeDecode.testEncodeLatin1          75  15       1.269 ±  0.028        1.247 ±  0.085  ms/op    1.8% (p = 0.310 )
  :gc.alloc.rate.norm                    N/A  15  807656.807 ±  0.191   524648.623 ±  0.589   B/op  -35.0% (p = 0.000*)
URLEncodeDecode.testEncodeLatin1         100  15       0.542 ±  0.000        0.542 ±  0.000  ms/op   -0.0% (p = 0.081 )
  :gc.alloc.rate.norm                    N/A  15       3.730 ±  0.004        3.732 ±  0.002   B/op    0.1% (p = 0.095 )
URLEncodeDecode.testEncodeUTF8             0  15       3.469 ±  0.137        2.850 ±  0.007  ms/op   17.8% (p = 0.000*)
  :gc.alloc.rate.norm                    N/A  15 3065136.041 ±  0.948  1982715.759 ±  0.072   B/op  -35.3% (p = 0.000*)
URLEncodeDecode.testEncodeUTF8            75  15       1.337 ±  0.031        1.249 ±  0.095  ms/op    6.5% (p = 0.002*)
  :gc.alloc.rate.norm                    N/A  15  821529.273 ±  0.216   529760.645 ±  0.657   B/op  -35.5% (p = 0.000*)
URLEncodeDecode.testEncodeUTF8           100  15       0.542 ±  0.000        0.542 ±  0.000  ms/op   -0.0% (p = 0.345 )
  :gc.alloc.rate.norm                    N/A  15       3.731 ±  0.001        3.730 ±  0.005   B/op   -0.0% (p = 0.698 )
  * = significant

Invariant parameters used by above microbenchmarks:
encodeChars:      6
maxLength  :   1024
```

dfuch commented on src/java.base/share/classes/java/net/URLEncoder.java:
```diff
@@ -239,7 +240,9 @@ public static String encode(String s, Charset charset) {
-        CharArrayWriter charArrayWriter = new CharArrayWriter();
+        CharsetEncoder ce = charset.newEncoder();
+        CharBuffer cb = CharBuffer.allocate(ENCODING_CHUNK_SIZE);
+        ByteBuffer bb = ByteBuffer.allocate((int)(ENCODING_CHUNK_SIZE * (double)ce.maxBytesPerChar()));
```
The cast to (double) is curious

cl4es replied:
It's copied from `String::scale`: 
```
    private static int scale(int len, float expansionFactor) {
        // We need to perform double, not float, arithmetic; otherwise
        // we lose low order bits when len is larger than 2**24.
        return (int)(len * (double)expansionFactor);
    }
```
... though admittedly this caveat doesn't apply since `len` in our case is a small constant value.

dfuch reviewed (approved):
Looks reasonable to me. Approved as long as tier2 still passes.

cl4es commented:
Thanks! I'll run tier1+tier2 before integrating.

RogerRiggs reviewed (approved):
Looks good, nice allocation reduction.

cl4es commented:
I hit a failing test in tier2 due missing REPLACE actions. Fixed that, also added a comment to clarify why I flush the buffer before it's full and added tests that would fail if surrogate pairs were split between chunks.

AlanBateman commented on src/java.base/share/classes/java/net/URLEncoder.java:
```diff
@@ -268,23 +275,47 @@ public static String encode(String s, Charset charset) {
         return out.toString();
     }
+
+    private static void flushToStringBuilder(StringBuilder out, CharsetEncoder ce, CharBuffer cb, ByteBuffer bb, boolean endOfInput) {
```
This method would benefit from a comment to summarise the parameters, only because there are so many. 

Also just a formatting nit is that this line is much longer than everything else in the file so maybe side-by-side diffs a bit difficult to look at.

cl4es replied:
Added a comment and split up the line.

AlanBateman commented on src/java.base/share/classes/java/net/URLEncoder.java:
```diff
@@ -239,11 +241,16 @@ public static String encode(String s, Charset charset) {
 
+        CharsetEncoder ce = charset.newEncoder()
+                .onMalformedInput(CodingErrorAction.REPLACE)
+                .onUnmappableCharacter(CodingErrorAction.REPLACE);
```
@dfuch The explicit use of the REPLACE action makes me wonder if the spec should be clarified at some point as it doesn't seem to be documented.

AlanBateman replied:
> I'd be happy to spell it out in the documentation, just pointing out that it's been long-standing behavior.

Right, I wasn't suggesting there is behavior change, instead I was pointing out to Daniel that the spec is incomplete and we should clarify it, it would be a separate update/PR and would require a CSR to track the spec update.

dfuch replied:
@AlanBateman I have logged [JDK-8316734](https://bugs.openjdk.org/browse/JDK-8316734)

ExE-Boss commented on src/java.base/share/classes/java/net/URLEncoder.java:
```diff
@@ -293,7 +293,23 @@ public static String encode(String s, Charset charset) {
 
-    private static void flushToStringBuilder(StringBuilder out, CharsetEncoder ce, CharBuffer cb, ByteBuffer bb, boolean endOfInput) {
+    /**
+     * Encodes input chars in cb and appends the byte values in an escaped format
+     * ("%FF") to out. The temporary byte buffer, bb, must be able to accept
+     * cb.position() * ce.maxBytesPerChar() bytes.
```
```suggestion
     * Encodes input chars in {@code cb} and appends the byte values in an escaped
     * format ({@code "%XX"}) to {@code out}. The temporary byte buffer, {@code bb},
     * must be able to accept {@code cb.position() * ce.maxBytesPerChar()} bytes.
```

cl4es replied:
As this javadoc isn't published anywhere I didn't bother with `{@code..`, but since you've spent time to fix things up then why not. 👍🏽

cl4es commented:
/integrate

wenshao commented:
> Micros show a small throughput win and a large allocation reduction for variants that need to change the URL either partially or completely, and no regression when the URL remains unchanged:
> 
> ```
> Name                             (unchanged) Cnt        Base    Error         Test    Error   Unit   Diff%
> URLEncodeDecode.testEncodeLatin1           0  15       3.471 ±  0.103        2.796 ±  0.078  ms/op   19.5% (p = 0.000*)
>   :gc.alloc.rate                                     828.462 ± 25.054      673.090 ± 19.214 MB/sec  -18.8% (p = 0.000*)
>   :gc.alloc.rate.norm                            3013680.062 ±  0.721  1972347.384 ±  0.540   B/op  -34.6% (p = 0.000*)
>   :gc.count                                           20.000                17.000          counts
>   :gc.time                                            16.000                15.000              ms
> URLEncodeDecode.testEncodeLatin1          75  15       1.269 ±  0.028        1.132 ±  0.029  ms/op   10.8% (p = 0.000*)
>   :gc.alloc.rate                                     606.924 ± 12.993      443.802 ± 11.184 MB/sec  -26.9% (p = 0.000*)
>   :gc.alloc.rate.norm                             807656.807 ±  0.191   526711.840 ±  0.197   B/op  -34.8% (p = 0.000*)
>   :gc.count                                           16.000                11.000          counts
>   :gc.time                                            16.000                10.000              ms
> URLEncodeDecode.testEncodeLatin1         100  15       0.542 ±  0.000        0.542 ±  0.000  ms/op   -0.0% (p = 0.932 )
>   :gc.alloc.rate                                       0.007 ±  0.000        0.007 ±  0.000 MB/sec    0.0% (p = 0.358 )
>   :gc.alloc.rate.norm                                  3.730 ±  0.004        3.731 ±  0.001   B/op    0.0% (p = 0.356 )
>   :gc.count                                            0.000                 0.000          counts
> URLEncodeDecode.testEncodeUTF8             0  15       3.469 ±  0.137        2.678 ±  0.023  ms/op   22.8% (p = 0.000*)
>   :gc.alloc.rate                                     843.593 ± 32.562      711.747 ±  6.147 MB/sec  -15.6% (p = 0.000*)
>   :gc.alloc.rate.norm                            3065136.041 ±  0.948  1999098.562 ±  0.161   B/op  -34.8% (p = 0.000*)
>   :gc.count                                           22.000                18.000          counts
>   :gc.time                                            20.000                15.000              ms
> URLEncodeDecode.testEncodeUTF8            75  15       1.337 ±  0.031        1.192 ±  0.107  ms/op   10.8% (p = 0.000*)
>   :gc.alloc.rate                                     586.133 ± 13.420      429.661 ± 36.595 MB/sec  -26.7% (p = 0.000*)
>   :gc.alloc.rate.norm                             821529.273 ±  0.216   533888.255 ±  0.744   B/op  -35.0% (p = 0.000*)
>   :gc.count                                           15.000                11.000          counts
>   :gc.time                                            10.000                10.000              ms
> URLEncodeDecode.testEncodeUTF8           100  15       0.542 ±  0.000        0.541 ±  0.000  ms/op    0.1% (p = 0.000*)
>   :gc.alloc.rate                                       0.007 ±  0.000        0.007 ±  0.000 MB/sec    0.0% (p = 0.771 )
>   :gc.alloc.rate.norm                                  3.731 ±  0.001        3.727 ±  0.002   B/op   -0.1% (p = 0.000*)
>   :gc.count                                            0.000                 0.000          counts
>   * = significant
> ```

Can you share how to generate this kind of diff report? What I am doing now is to run the baseline and the current version separately, and then use the processing program I wrote to generate the diff report. Can you share your experience?

cl4es commented:
> > Micros show a small throughput win and a large allocation reduction for variants that need to change the URL either partially or completely, and no regression when the URL remains unchanged:
> > ```
> > Name                             (unchanged) Cnt        Base    Error         Test    Error   Unit   Diff%
> > URLEncodeDecode.testEncodeLatin1           0  15       3.471 ±  0.103        2.796 ±  0.078  ms/op   19.5% (p = 0.000*)
> >   :gc.alloc.rate                                     828.462 ± 25.054      673.090 ± 19.214 MB/sec  -18.8% (p = 0.000*)
> >   :gc.alloc.rate.norm                            3013680.062 ±  0.721  1972347.384 ±  0.540   B/op  -34.6% (p = 0.000*)
> >   :gc.count                                           20.000                17.000          counts
> >   :gc.time                                            16.000                15.000              ms
> > <snip>
> > ```
> 
> Can you share how to generate this kind of diff report? What I am doing now is to run the baseline and the current version separately, and then use the processing program I wrote to generate the diff report. Can you share your experience?

It's a small utility I've toyed with and iterated on over the last couple of weeks. It's tangled up on some legacy proprietary code, but I think I can refactor out the relevant pieces and put it in my OpenJDK repo in the next couple of weeks.

wenshao commented:
URLEncoder#DONT_NEED_ENCODING based on BitSet is actually a lookup table. Should we consider improving it in this way?

```java
public class URLEncoder {
	static final long DONT_NEED_ENCODING_FLAGS_0;
	static final long DONT_NEED_ENCODING_FLAGS_1;

	static {
		long flag0 = 0;
	    flag0 |= 1L << ' '; // ASCII 32
	    flag0 |= 1L << '*'; // ASCII 42
	    flag0 |= 1L << '-'; // ASCII 25
	    flag0 |= 1L << '.'; // ASCII 46

	    // ASCII 48 - 57
	    for (int i = '0'; i <= '9'; ++i) {
	        flag0 |= 1L << i;
	    }
	    DONT_NEED_ENCODING_FLAGS_0 = flag0;

	    long flags1 = 0;
	    // ASCII 65 - 90
	    for (int i = 'A'; i <= 'Z'; ++i) {
	        flags1 |= 1L << (i - 64);
	    }
	    flags1 |= 1L << ('_' - 64); // ASCII 95
	    // ASCII 97 - 122
	    for (int i = 'a'; i <= 'z'; ++i) {
	        flags1 |= 1L << (i - 64);
	    }
	    DONT_NEED_ENCODING_FLAGS_1 = flags1;
	}

    private static boolean dontNeedEncoding(char c) {
        int prefix = c >> 6;
        if (prefix > 1) {
            return false;
        }
        long flags = prefix == 0 ? DONT_NEED_ENCODING_FLAGS_0 : DONT_NEED_ENCODING_FLAGS_1;
        return (flags & (1L << c)) != 0;
    }
}
```

cl4es commented:
How would this be an improvement?

wenshao commented:
The improvement is about 3% to 13% under microbench. The baseline is based on BitSet table lookup. In real scenarios, the performance of the new version that does not use BitSet should be better.

Performance numbers running on MacBook M1 Pro: 
```diff
-Benchmark                         (encodeChars)  (maxLength)  (unchanged)  Mode  Cnt  Score   Error  Units (baseline)
-URLEncodeDecode.testEncodeLatin1              6         1024            0  avgt   15  2.244 ? 0.290  ms/op
-URLEncodeDecode.testEncodeLatin1              6         1024           75  avgt   15  0.956 ? 0.074  ms/op
-URLEncodeDecode.testEncodeLatin1              6         1024          100  avgt   15  0.400 ? 0.006  ms/op

+Benchmark                         (encodeChars)  (maxLength)  (unchanged)  Mode  Cnt  Score    Error  Units (optimized)
+URLEncodeDecode.testEncodeLatin1              6         1024            0  avgt   15  2.057 ?  0.016  ms/op (+9.10)
+URLEncodeDecode.testEncodeLatin1              6         1024           75  avgt   15  0.842 ?  0.002  ms/op (+13.54)
+URLEncodeDecode.testEncodeLatin1              6         1024          100  avgt   15  0.385 ?  0.001  ms/op (+3.90)
```
